### ISSUE TITLE
Oddity in processing of indented string

### ISSUE THREAD

**Issue by ppslim** [pdvrieze/xmlutil#20]
I have having some dfficulty trying to understand why problems are being seen. Though the core issue looks to be in `StAXReader` and above.

Below is a very quick sample after dragging down my diagnostics to the most basic level I can find.

```kotlin
class SampleParser {
    val baseXmlFormat = XML

    fun getSampleEntity(): TestBaseModel {
        val serializer = TestBaseModel.serializer()
        val input = getSampelXmlImplA()
        println(input)
        return baseXmlFormat.parse(serializer, input)
    }

    fun getSampelXmlImplA(): String {
        return """
            <replicatedDirectory>
            <path>/a/b/c</path>
            <name>repl1</name>
            <gsn>1024</gsn>
        </replicatedDirectory>
        """.trimIndent()
    }

    fun getSampelXmlImplB(): String {
        return """
        <replicatedDirectory>
            <path>/a/b/c</path>
            <name>repl1</name>
            <gsn>1024</gsn>
        </replicatedDirectory>
        """.trimIndent()
    }
}
```

When using the XML string returned by `getSampelXmlImplA`, parsing of my data is perfectly fine.

However, if parsing the XML string returned by `getSampelXmlImplB`, parsing dives into an indefinite loop.

A snapshot of the parsing thread.

```
"main" #1 prio=5 os_prio=0 tid=0x00007f4ef800d000 nid=0x6299 runnable [0x00007f4efebeb000]
   java.lang.Thread.State: RUNNABLE
	at com.sun.org.apache.xerces.internal.impl.XMLEntityScanner.arrangeCapacity(XMLEntityScanner.java:1764)
	at com.sun.org.apache.xerces.internal.impl.XMLEntityScanner.skipString(XMLEntityScanner.java:1799)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanEndElement(XMLDocumentFragmentScannerImpl.java:1748)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2967)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:602)
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
	at com.sun.org.apache.xerces.internal.impl.XMLStreamReaderImpl.next(XMLStreamReaderImpl.java:553)
	at nl.adaptivity.xmlutil.StAXReader.next(StAXReader.kt:172)
	at nl.adaptivity.xmlutil.XmlReaderUtil__XmlReaderKt.readSimpleElement(XmlReader.kt:278)
	at nl.adaptivity.xmlutil.XmlReaderUtil.readSimpleElement(Unknown Source)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$TagDecoder.decodeStringElement(XMLDecoder.kt:777)
	at com.github.ppslim.models.samples.TestBaseModel$$serializer.deserialize(TestBaseModel.kt)
	at com.github.ppslim.models.samples.TestBaseModel$$serializer.deserialize(TestBaseModel.kt:8)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$XmlDecoder.decodeSerializableValue(XMLDecoder.kt:140)
	at nl.adaptivity.xmlutil.serialization.XML.parse(XML.kt:333)
	at nl.adaptivity.xmlutil.serialization.XML.parse(XML.kt:351)
	at nl.adaptivity.xmlutil.serialization.XML$Companion.parse(XML.kt:475)
	at com.github.ppslim.models.samples.SampleModel.getSampleEntity(SampleModel.kt:22)
	at com.github.ppslim.models.samples.SampleModelKt.main(SampleModel.kt:10)
	at com.github.ppslim.models.samples.SampleModelKt.main(SampleModel.kt)
```
Adding some breakpoints to `XMLEntityScanner.arrangeCapacity`, I can see that it never breaks out of the while loop.

I doubt the problem is in this library, but it has been confusing me all day

**Comment by pdvrieze** [comment]
This is a weird one. There are various known limitations in the XML parser bundled in the JDK. In my own projects I use woodstox for this reason. Anyway I'll have a look at it. my StAXReader class just forwards to the underlying stream. Clear is that the error happens somewhere while parsing one of the elements with text content ("path", "gsn" or "name").

**Comment by ppslim** [comment]
I suspect this is a knowledge issue, but I was not fully aware that Stax had a SPI I could substitute in place. In the short term, Woodstox is as least allowing me to progress.

Given this looks to be problem related to the default SPI in the JVM, so it does limit what this project can do. Though I suspect it would be worthy of a note, especially around the JVM elements of "how to use"

**Comment by pdvrieze** [comment]
I believe I've found the problem. It is in a class I wrote `CharsequenceReader` where reading into a buffer at an offset was done incorrectly (wrong math that happens to work if the offset is 0). This has now been fixed. Woodstox happens to read things in differently and not hit the bug (and is used for testing as it generates better xml). Fixed in: a7923ea40a8ac3554467911693f0d889fd6e4ff0 and soon to-be-released bugfix release 0.13.0.1.
